word_segmentation.py

```diff
@@ -0,0 +1,709 @@
+import numpy as np
+import os
+import icu
+import datetime
+from icu import UnicodeString, BreakIterator, Locale
+import matplotlib.pyplot as plt
+
+import tensorflow as tf
+from keras.models import Sequential
+from keras.layers import LSTM
+from keras.layers import Dense
+from keras.layers import TimeDistributed
+from keras.layers import Bidirectional
+from keras.layers import Embedding
+from keras.layers import Dropout
+
+
+def is_all_english(str):
+    """
+    A very basic function that checks if all elements of str are english characters or common expressions
+    It just checks if the integer assigned to each character is below 130 or not
+    TO DO: This function can be potentially be replace with some ICU function
+    str: input string
+    """
+
+    str = str.replace(" ", "")
+    str = str.replace("|", "")
+    # for k in range(0, 130):
+    #     print(chr(k))
+    for ch in str:
+        if ord(ch) > 130:
+            # print("character = {}, ord = {}".format(ch, ord(ch)))
+            return False
+    return True
+
+
+def sigmoid(x):
+    """
+    Computes the sigmoid function for a scalar
+    x: the scalar
+    """
+    return 1.0/(1.0+np.exp(-x))
+
+
+def analyze_grapheme_clusters(ratios, thrsh):
+    """
+    This function analyzes the grapheme clusters, to see what percentage of them form which percent of the text, and
+    provides a histogram that shows frequency of grapheme clusters
+    ratios: a dictionary that holds the ratio of text that is represented by each grapheme cluster
+    thrsh: shows what percent of the text we want to cover
+    """
+    cum_sum = 0
+    cnt = 0
+    for val in ratios.values():
+        cum_sum += val
+        cnt += 1
+        if cum_sum > thrsh:
+            break
+    print("number of different grapheme clusters = {}".format(len(ratios.keys())))
+    print("{} grapheme clusters form {} of the text".format(cnt, thrsh))
+    # plt.hist(ratios.values(), bins=50)
+    # plt.show()
+
+
+def get_segmented_string(str, brkpoints):
+    """
+    Rerurns a segmented string using the unsegmented string and and break points. Simply inserts | at break points.
+    str: unsegmented string
+    brkpoints: break points
+    """
+    out = "|"
+    for i in range(len(brkpoints)-1):
+        out += str[brkpoints[i]: brkpoints[i+1]] + "|"
+    return out
+
+
+def get_bies(char_brkpoints, word_brkpoints):
+    """
+    Given break points for words and grapheme clusters, returns the matrix that represents BIES.
+    The output is a matrix of size (n * 4) where n is the number of grapheme clusters in the string
+    char_brkpoints: break points for grapheme clusters
+    word_brkpoints: break points for words
+    """
+    bies = np.zeros(shape=[4, len(char_brkpoints)-1])
+    word_ind = 0
+    for i in range(len(char_brkpoints)-1):
+        word_st = word_brkpoints[word_ind]
+        word_fn = word_brkpoints[word_ind + 1]
```

Comment:
I don't want you to change it in this file, but in the future, consider using "start" and "limit" or even just "0" and "1" for begin and end variable names.  I assume "st" means "start" and "fn" means "finish", but when I see "fn" it looks more like "function" to me.

Reply:
The reason that I don't just use "start" and "finish" is that many times I am working with begin and end of more than one thing (like characters and words here). I think the best thing is that I just use a full name that explains what is that variable, such as "word_start" and "word_finish".